packages/react-reconciler/src/ReactChildFiber.js

```diff
@@ -1342,9 +1383,140 @@ function ChildReconciler(shouldClone, shouldTrackSideEffects) {
     return created;
   }
 
+  // A fork of reconcileChildFibers to be used when reconcileChildFibers
+  // encounters a top level fragment to treat it as a set of children.
+  // This function is not recursive.
+  function reconcileChildFibersForTopLevelFragment(
+    returnFiber: Fiber,
+    currentFirstChild: Fiber | null,
+    newChild: any,
+    expirationTime: ExpirationTime,
+  ): Fiber | null {
+    // This function is not recursive.
+    // If the top level item is an array, we treat it as a set of children,
+    // not as a fragment. Nested arrays on the other hand will be treated as
+    // fragment nodes. Recursion happens at the normal flow.
+
+    // Handle object types
+    const isObject = typeof newChild === 'object' && newChild !== null;
+    if (isObject) {
+      switch (newChild.$$typeof) {
+        case REACT_ELEMENT_TYPE:
+          return placeSingleChild(
+            reconcileSingleElement(
+              returnFiber,
+              currentFirstChild,
+              newChild,
+              expirationTime,
+            ),
+          );
+        case REACT_COROUTINE_TYPE:
+          return placeSingleChild(
+            reconcileSingleCoroutine(
+              returnFiber,
+              currentFirstChild,
+              newChild,
+              expirationTime,
+            ),
+          );
+        case REACT_YIELD_TYPE:
+          return placeSingleChild(
+            reconcileSingleYield(
+              returnFiber,
+              currentFirstChild,
+              newChild,
+              expirationTime,
+            ),
+          );
+        case REACT_PORTAL_TYPE:
+          return placeSingleChild(
+            reconcileSinglePortal(
+              returnFiber,
+              currentFirstChild,
+              newChild,
+              expirationTime,
+            ),
+          );
+      }
+    }
+
+    if (typeof newChild === 'string' || typeof newChild === 'number') {
+      return placeSingleChild(
+        reconcileSingleTextNode(
+          returnFiber,
+          currentFirstChild,
+          '' + newChild,
+          expirationTime,
+        ),
+      );
+    }
+
+    if (isArray(newChild)) {
```

Comment:
Doesn't this mean that `return <>{[...]}</>` is treated the same as `return <>{...}</>`. That's not the same as `return [...]` vs `return [[...]]`. It flattens an extra level because reconcileChildrenArray avoids the extra fragment but in this case there should always be a fragment.

We should probably have a test for this.

Reply:
I think I gave you bad advice. The function you wanted to fork was `updateSlot`.

Reply:
Bringing this up made me think about some other permutations of fragment and array combinations given our current desired behavior of top-level fragments behaving the same as a top-level array. I jotted some of them down here (with the ones I'm unsure about marked with `?`:
https://gist.github.com/clemmy/b3ef00f9507909429d8aa0d3ee4f986b
Does that look right to you?